### ISSUE TITLE
Generating Parser for C#

### ISSUE THREAD

**Issue by ericlemes** [tunnelvisionlabs/antlr4cs#263]
Hi there,

Thank you very much for the great work here. 

I'm currently trying to generate a C++ Parser using Antlr, to run in C#. I'm using VS2017. I've bumped into a lot of memory, then performance issues, which I'm trying to work around, and that's the reason I'm trying this C# target.

I've referenced the latest packages from NuGet (Antlr4, Antlr4.CodeGenerator and Antlr4.RunTime), all Sam Harwell, all 4.6.4.

I'm generating the code from command line using the tool:

    "C:\Program Files\Java\jre1.8.0_151\bin\java.exe" -jar E:\SandBox\AntlrCppParser\build\..\src\packages\Antlr4.CodeGenerator.4.6.4\tools\antlr4-csharp-4.6.4-complete.jar E:\SandBox\AntlrCppParser\build\..\lib\CPP14.g4 -o E:\SandBox\Antl  lrCppParser\build\..\src\AntlrCppParser.AntlrParser -Dlanguage=CSharp_v4_5

The grammar is the C++ grammar in antlr grammar repo (https://github.com/antlr/grammars-v4/tree/master/cpp).

When I wrap the generated files in a C#, .NET 4.5.2 Project, I get the following compilation error:

1>E:\SandBox\AntlrCppParser\src\AntlrCppParser.AntlrParser\CPP14Lexer.cs(164,27,164,37): error CS0115: 'CPP14Lexer.TokenNames': no suitable method found to override
1>E:\SandBox\AntlrCppParser\src\AntlrCppParser.AntlrParser\CPP14Parser.cs(245,27,245,37): error CS0115: 'CPP14Parser.TokenNames': no suitable method found to override

Any thoughts?

Cheers,

Eric

**Comment by KvanTTT** [comment]
Are you using a [Standard](https://github.com/antlr/antlr4/tree/4.7/runtime/CSharp) runtime or this runtime by @sharwell?

**Comment by ericlemes** [comment]
I'm using @sharwell's runtime. At least this is my intention. It is the Antlr.Runtime 4.6.4, nuget package, by Sam Harwell.

**Comment by KvanTTT** [comment]
In this case, you have to use both Sam parser generator and runtime itself.

**Comment by ericlemes** [comment]
@KvanTTT, all the referenced packages are from @sharwell . The antlr tools I've used by command line is also the same version as used in the deployed nuget packages

**Comment by sharwell** [comment]
> I'm generating the code from command line using the tool ...

This usage is not supported. This target automatically generated code during the build process, which ensures that the generated code matches the grammar and the generated code matches the runtime library.

**Comment by andrewsav-datacom** [comment]
@sharwell does this mean that Visual Studio is a requirement and there is no way to use this with other dotnet IDEs?

**Comment by lextm** [comment]
@andrewsav-datacom Have you tried? So far all major .NET IDE products use MSBuild as build engine, and they (including Rider, and VS for Mac) work well with ANTLR.

**Comment by AndrewSav** [comment]
@lextm yes, cannot get it to work with VS Code, getting the error as in OP and https://marketplace.visualstudio.com/items?itemName=mike-lischke.vscode-antlr4

**Comment by AndrewSav** [comment]
May be version matrix can help, so that users know which version of runtime correspond to which version of official antlr tools. I'm usually able to figure out most issues like this on my own, but in this case I told myself, okay, let me troubleshoot, let me exclude the middleman (VSCode pluging) and go to the basics, let's generate with antlr and reference the execution engine. I tried about half a dozen combinations of antlr and the runtime versions, and all of them gave me errors, different ones too.

So yes, I have tried, spend about a day on it. 

I would be also awesome to find out _why_ it's not supported, that is, what kind of magic msbuild process (roughly) does, that following official antlr4 java documentation/book does not have.

I just looked over code in this repository and it appears that it corresponds to version 4.6 of antlr4 but could not get it to work in this combination either.  Just trying to figure out what's missing, how is this working under the hood. Any help?

**Comment by lextm** [comment]
@AndrewSav You should report any issue with a VSCode extension to its authors, and maybe put a link here. It would be much sooner for them to figure out what's wrong, than bothering yourself.

ANTLR does not use too much MSBuild magic, as even command line build should work flawlessly. But I don't assume ANTLR a small project to get started all by yourself.

**Comment by AndrewSav** [comment]
@sharwell cool, thanks, gotcha. My mistake was that I assumed that this is the repo with the official runtime. I knew that your code generation was different from official, but I failed to realize that runtimes where different too. I thought that both code generation targets work with the same one and only runtime, that is hosted in this repo. Now, when I realized that it's not, everything else fell into place.
